server/accounts_test.go

```diff
@@ -1406,6 +1406,139 @@ func TestAccountRequestReplyTrackLatency(t *testing.T) {
+func TestAccountTrackLatencyRemoteLeaks(t *testing.T) {
+	optsA, _ := ProcessConfigFile("./configs/seed.conf")
+	optsA.NoSigs, optsA.NoLog = true, true
+	srvA := RunServer(optsA)
+	defer srvA.Shutdown()
+	optsB := nextServerOpts(optsA)
+	optsB.Routes = RoutesFromStr(fmt.Sprintf("nats://%s:%d", optsA.Cluster.Host, optsA.Cluster.Port))
+	srvB := RunServer(optsB)
+	defer srvB.Shutdown()
+
+	checkClusterFormed(t, srvA, srvB)
+	srvs := []*Server{srvA, srvB}
+
+	// Now add in the accounts and setup tracking.
+	for _, s := range srvs {
+		s.SetSystemAccount(globalAccountName)
+		fooAcc, _ := s.RegisterAccount("$foo")
+		fooAcc.AddServiceExport("track.service", nil)
+		fooAcc.TrackServiceExport("track.service", "results")
+		barAcc, _ := s.RegisterAccount("$bar")
+		if err := barAcc.AddServiceImport(fooAcc, "req", "track.service"); err != nil {
+			t.Fatalf("Failed to import: %v", err)
+		}
+	}
+
+	// Test with a responder on second server, srvB. but they will not respond.
+	cfoo, crFoo, _ := newClientForServer(srvB)
+	defer cfoo.nc.Close()
+	fooAcc, _ := srvB.LookupAccount("$foo")
+	if err := cfoo.registerWithAccount(fooAcc); err != nil {
+		t.Fatalf("Error registering client with 'foo' account: %v", err)
+	}
+
+	// Set new limits
+	fooAcc.SetAutoExpireTTL(time.Millisecond)
+	fooAcc.SetMaxAutoExpireResponseMaps(5)
+
+	// Now setup the resonder under cfoo and the listener for the results
+	baseSubs := int(srvA.NumSubscriptions())
+	cfoo.parse([]byte("SUB track.service 1\r\n"))
+	// Wait for it to propagate.
+	checkExpectedSubs(t, baseSubs+1, srvA)
+
+	cbar, _, _ := newClientForServer(srvA)
+	defer cbar.nc.Close()
+	barAcc, _ := srvA.LookupAccount("$bar")
+	if err := cbar.registerWithAccount(barAcc); err != nil {
+		t.Fatalf("Error registering client with 'bar' account: %v", err)
+	}
+
+	// Send 2 requests
+	go cbar.parseAndFlush([]byte("SUB resp 11\r\nPUB req resp 4\r\nhelp\r\nPUB req resp 4\r\nhelp\r\n"))
+
+	readFooMsg := func() ([]byte, string) {
+		t.Helper()
+		l, err := crFoo.ReadString('\n')
+		if err != nil {
+			t.Fatalf("Error reading from client 'bar': %v", err)
+		}
+		mraw := msgPat.FindAllStringSubmatch(l, -1)
+		if len(mraw) == 0 {
+			t.Fatalf("No message received")
+		}
+		msg := mraw[0]
+		msgSize, _ := strconv.Atoi(msg[LEN_INDEX])
+		return grabPayload(crFoo, msgSize), msg[REPLY_INDEX]
+	}
+
+	readFooMsg()
+	readFooMsg()
+
+	var rc *client
+	// Pull out first client
+	srvB.mu.Lock()
+	for _, rc = range srvB.clients {
+		if rc != nil {
+			break
+		}
+	}
+	srvB.mu.Unlock()
+
+	tracking := func() int {
+		rc.mu.Lock()
+		numTracking := len(rc.rrTracking)
+		rc.mu.Unlock()
+		return numTracking
+	}
+
+	numTracking := tracking()
+
+	if numTracking != 2 {
+		t.Fatalf("Expected to have 2 tracking replies, got %d", numTracking)
+	}
+
+	// Make sure these remote tracking replies honor the current auto expire TTL.
+	time.Sleep(2 * time.Millisecond)
+
+	rc.mu.Lock()
```

Comment:
This is really strange. You invoke the pruning manually here? I don't see a place where we remove those entries from rrTracking simply based on TTL (we do based on count).

Reply:
Correct, same as other auto-expire, we remove when there is pressure and we hit a threshold. I need to rework but wanted to make this consistent with the others for now.